contracts/governance/src/ZeroExVotes.sol

```diff
@@ -0,0 +1,282 @@
+        );
+
+        emit TotalSupplyChanged(newWeight, newQuadraticWeight);
+    }
+
+    /**
+     * @inheritdoc IZeroExVotes
+     */
+    function writeCheckpointTotalSupplyBurn(uint256 accountBalance, uint256 amount) public override onlyToken {
+        (, uint256 newWeight, , uint256 newQuadraticWeight) = _writeCheckpoint(
+            _totalSupplyCheckpoints,
+            _subtract,
+            accountBalance,
+            amount
+        );
+
+        emit TotalSupplyChanged(newWeight, newQuadraticWeight);
+    }
+
+    /**
+     * @dev Lookup a value in a list of (sorted) checkpoints.
+     * Implementation as in openzeppelin/token/ERC20/extensions/ERC20Votes.sol except here we return the entire
+     * checkpoint rather than part of it
+     */
+    function _checkpointsLookup(
+        Checkpoint[] storage ckpts,
+        uint256 blockNumber
+    ) private view returns (Checkpoint memory) {
+        // We run a binary search to look for the earliest checkpoint taken after `blockNumber`.
+        //
+        // Initially we check if the block is recent to narrow the search range.
+        // During the loop, the index of the wanted checkpoint remains in the range [low-1, high).
+        // With each iteration, either `low` or `high` is moved towards the middle of the range to maintain the
+        // invariant.
+        // - If the middle checkpoint is after `blockNumber`, we look in [low, mid)
+        // - If the middle checkpoint is before or equal to `blockNumber`, we look in [mid+1, high)
+        // Once we reach a single value (when low == high), we've found the right checkpoint at the index high-1, if not
+        // out of bounds (in which case we're looking too far in the past and the result is 0).
+        // Note that if the latest checkpoint available is exactly for `blockNumber`, we end up with an index that is
+        // past the end of the array, so we technically don't find a checkpoint after `blockNumber`, but it works out
+        // the same.
+        uint256 length = ckpts.length;
+
+        uint256 low = 0;
+        uint256 high = length;
+
+        if (length > 5) {
+            uint256 mid = length - Math.sqrt(length);
+            if (_unsafeAccess(ckpts, mid).fromBlock > blockNumber) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        while (low < high) {
+            uint256 mid = Math.average(low, high);
+            if (_unsafeAccess(ckpts, mid).fromBlock > blockNumber) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        // Leaving here for posterity this is the original OZ implementation which we've replaced
+        // return high == 0 ? 0 : _unsafeAccess(ckpts, high - 1).votes;
+        Checkpoint memory checkpoint = high == 0 ? Checkpoint(0, 0, 0) : _unsafeAccess(ckpts, high - 1);
+        return checkpoint;
+    }
+
+    function _writeCheckpoint(
```

Comment:
Can we make more liberal use of `unchecked` throughout this function? Seems like none of the math here can overflow. Likewise, we ought to be able to remove the use of `SafeCast`